File: project_euler/01/019.go
// Copyright [2017] <dmnsn7@gmail.com>

package main

import "fmt"

const yearBegin = 1900
const yearEnd = 2000

var monthDay = []int{31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31}

func isLeap(year int) bool {
	return year%400 == 0 || year%100 != 0 && year%4 == 0
}

func main() {
	year, month, week := yearBegin, 0, 1
	cnt := 0
	for year <= yearEnd {
		if year >= yearBegin+1 && week == 0 {
			cnt++
		}
		if month == 1 && isLeap(year) {
			week = (week + monthDay[month] + 1) % 7
		} else {
			week = (week + monthDay[month]) % 7
		}
		if month == 11 {
			year++
			month = 0
		} else {
			month++
		}
	}

	fmt.Printf("%d\n", cnt)
}

File: project_euler/01/019.py
#!/usr/bin/env python3
'''Copyright [2017] <dmnsn7@gmail.com>'''

YEAR_BEGIN = 1900
YEAR_END = 2000

MONTH_DAY = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]


def is_leap(year):
    return year % 400 == 0 or year % 100 != 0 and year % 4 == 0


def main():
    year, month, week = YEAR_BEGIN, 0, 1
    cnt = 0
    while year <= YEAR_END:
        cnt += 1 if year >= YEAR_BEGIN + 1 and week == 0 else 0
        week = (week + MONTH_DAY[month] + (month == 1 and is_leap(year))) % 7
        year += 1 if month == 11 else 0
        month = 0 if month == 11 else month + 1

    print(cnt)


if __name__ == '__main__':
    main()
